nixos/modules/virtualisation/qemu-vm.nix

```diff
@@ -558,10 +560,22 @@ in
       };
 
     virtualisation.qemu = {
+      pkgs = mkOption {
```

Comment:
This is not a property of qemu, but of the host. I'd suggest `virtualisation.host.pkgs` instead of `virtualisation.qemu.pkgs`. It matches the description better, and also aligns with `hostPkgs` in https://github.com/NixOS/nixpkgs/pull/176557

-----

_Thoughts_

The homonymy with `nixpkgs.hostPlatform` is unfortunate, but I don't think we can use another term than `host`. `hypervisor` is too specific (it would be nice for the option to be usable in a container-based setting). `monitor` is just worse.
At least the `virtualisation` prefix gives a clue to resolve the ambiguity, so it seems that `virtualisation.host.pkgs` is the best we can do.

(As a side note, arguably the host isn't part of the system, but it's a little late to backtrack on the `virtualisation.qemu.package` option, which is also a host property.)

Reply:
I'm not sure we should break consistency with `virtualisation.qemu.package` here. `virtualisation` namespace as it is is quite a mess of both "this config as a host" and "this config as a guest" options. I don't think adding another one directly there would be better than keeping everything related to qemu-vm module to `virtualisation.qemu`. Besides, this package set is used only for building and wrapping the qemu binary itself.

If we want to accommodate possibility of building other artefacts from the config (like Docker image, maybe?), I think we should have this option somewhere in a more generic place like `system.build.host.pkgs` (although I don't think we can add anything to `system.build`). I think this would be a part of a separate effort to unify such abstraction across all builders, which is not the aim of this pull request.

Reply:
Why wait? You'd just create more legacy options.

> I think we should have this option somewhere in a more generic place like `system.build.host.pkgs`

`system.build` is for the actual "outputs" of a config, such as `toplevel` and images.

`host.pkgs` is more of an input, like most options, including `nixpkgs.pkgs`. It just happens to be used in derivations that may produce outputs in `system.build`.

> (although I don't think we can add anything to `system.build`).

I don't think there's a technical limitation. You can even define proper options there nowadays, instead of just config attrs.

Reply:
I actually didn't know about `nixpkgs.pkgs` at all. It looks like we already have `hostPlatform` and `buildPlatform` there. Maybe we should also have `nixpkgs.targetPlatform` and `nixpkgs.targetPkgs` or something like this? It would keep the terminology consistent with cross-compilation.

Then if I'm on aarch64-darwin laptop want to cross-compile aarch64-linux system on x86_64-linux and run it in VM, I will have:

```
nixpkgs.hostPlatform.system = "aarch64-linux";
nixpkgs.buildPlatform.system = "x86_64-linux";
nixpkgs.targetPlatform.system = "aarch64-darwin";
```

And then we can set `nixpkgs.targetPlatform.system` automatically by `nixos-rebuild build-vm`, for example, and use it for any future additions like `nixos-rebuild build-docker-image`.

I feel like `nixpkgs` would be a better place for this option (or options) than `virtualisation`, but I'm not sure about the naming. The meaning of these terms in cross-compiling space always confuses me.

Reply:
I agree with @roberth on this, and I think his original proposal of `virtualisation.host.pkgs` is a good idea, since it's specific to only virtualisation. We could also rename `virtualisation.qemu.package` to `virtualisation.qemu.host.package` if we want all host options a `host` in their option path.

Reply:
Apologies if the first two parts of my reply are redundant (because you seem to agree) but some things should be clarified.

> I didn't mean that `targetPlatform` would be used for stdenv configuration. As you correctly stated, this would make the build result depend on which platform evaluation happens.

I don't think that's what I've stated. Perhaps I should have stressed that the "evaluation platform" is forgotten after instantiation and does not even exist at all in pure mode.

> I didn't mean that `targetPlatform` would be used for stdenv configuration.

Then let's avoid this name. Application of the original `targetPlatform` concept in `stdenv` is confusing enough as it is, so let's not add any ambiguity. A VM is usually not a compiler, and if it is, it is a JIT, so `build`, `host` and `target` are the same. The only thing that isn't the same is the source architecture.

Were you referring to `nixos-rebuild --target-host`? That option has basically nothing to do with `targetPlatform` and should be renamed. It just so happens that _during Nixpkgs bootstrapping_ these terms do line up somewhat, but bootstrapping is abstracted away by Nixpkgs and is not something you should ever have to worry about, unless you're packaging compilers that are part of stdenv.

> That's why I'm suggesting that `nixos-rebuild build-vm` could provide local system to the NixOS configuration in `virtualisation.host.pkgs` or `virtualisation.host.platform.system`, so that the resulting script can run locally on the system that `nixes-rebuild` has been called on.

This is definitely something we should pay attention to. Essentially you're asking for `virtualisation.host` to become impure. This is straightforward in the non-flakes case, and otherwise feasible by making `nixos-rebuild` invoke `extendModules` on the configuration, although I'm not sure if we should.
I don't think we have to implement this functionality for this PR to be useful, so I'd recommend to follow it up later.